File: js/examples/larry.ts
/**
 * Larry Williams 식 변동성 돌파 전략 (롱 온리, 다중 종목).
 * 직전 완성 캔들 몸통 >= 평균 몸통 x multiplier 인 양봉 -> 시장가 매수 (손절=시가, 브라켓 위임).
 *
 * 실행: HERMETIX_BROKER=next NEXT_CLIENT_ID=... node dist/examples/larry.js
 */
import { Decimal } from "decimal.js";
import type { CandleInterval, Signal, Strategy, StrategyContext, StrategySpec } from "../src/index.js";

export interface LarryOptions {
  symbols?: string[];
  lookback?: number;
  multiplier?: Decimal;
  expireHours?: number;
  budgetRatio?: Decimal;
  candleInterval?: CandleInterval;
}

export class LarryStrategy implements Strategy {
  readonly spec: StrategySpec;
  private readonly lookback: number;
  private readonly multiplier: Decimal;
  private readonly expireHours: number;
  private readonly budgetRatio: Decimal;
  readonly lastEvaluated = new Map<string, number>();
  readonly entryAt = new Map<string, number>();

  constructor(opts: LarryOptions = {}) {
    const symbols = opts.symbols ?? ["AAPL"];
    this.lookback = opts.lookback ?? 24;
    this.multiplier = opts.multiplier ?? new Decimal("1.2");
    this.expireHours = opts.expireHours ?? 48;
    this.budgetRatio = opts.budgetRatio ?? new Decimal("0.5");
    this.spec = {
      name: "larry", symbols,
      candleInterval: opts.candleInterval ?? "1h",
      candleLimit: this.lookback + 2,
    };
  }

  decide(ctx: StrategyContext): Signal[] {
    return this.spec.symbols.flatMap((symbol) => this.decideSymbol(symbol, ctx));
  }

  private decideSymbol(symbol: string, ctx: StrategyContext): Signal[] {
    if (ctx.hasPosition(symbol)) return this.decideExit(symbol, ctx);
    this.entryAt.delete(symbol);
    if (ctx.hasOpenOrder(symbol)) return [];
    return this.decideEntry(symbol, ctx);
  }

  private decideEntry(symbol: string, ctx: StrategyContext): Signal[] {
    const candles = ctx.candlesOf(symbol);
    if (candles.length < this.lookback + 2) return [];
    const completed = candles.slice(0, -1); // 마지막 캔들은 진행 중
    const target = completed[completed.length - 1];
    if (this.lastEvaluated.get(symbol) === target.timestamp.getTime()) return [];
    this.lastEvaluated.set(symbol, target.timestamp.getTime());

    const history = completed.slice(-(this.lookback + 1), -1);
    const avgBody = history
      .reduce((acc, c) => acc.plus(c.open.minus(c.close).abs()), new Decimal(0))
      .div(history.length);
    if (target.open.minus(target.close).abs().lt(avgBody.mul(this.multiplier))) return [];
    if (target.open.gte(target.close)) return []; // 롱 온리 - 음봉 스킵

    const quote = ctx.quote(symbol);
    if (!quote) return [];
    const budget = ctx.buyingPower.mul(this.budgetRatio).div(this.spec.symbols.length);
    const quantity = budget.div(quote.price).floor();
    if (quantity.lt(1)) return [];

    this.entryAt.set(symbol, ctx.now.getTime());
    return [{ kind: "buy", symbol, quantity, stopLossPrice: target.open }];
  }

  private decideExit(symbol: string, ctx: StrategyContext): Signal[] {
    const openedAt = this.entryAt.get(symbol) ?? ctx.now.getTime();
    this.entryAt.set(symbol, openedAt); // 재시작 시 만료 클록 재시작
    if (ctx.now.getTime() - openedAt < this.expireHours * 3600_000) return [];
    const holding = ctx.holding(symbol);
    if (!holding) return [];
    this.entryAt.delete(symbol);
    return [{ kind: "sell", symbol, quantity: holding.quantity }];
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const { run } = await import("./_runner.js");
  await run(new LarryStrategy());
}

File: python/examples/larry.py
"""Larry Williams 식 변동성 돌파 전략 (롱 온리, 다중 종목).

직전 완성 캔들의 몸통이 평균 몸통 x multiplier 이상인 양봉이면 시장가 매수,
손절은 진입 캔들 시가(브라켓 위임), expire_hours 경과 시 만료 청산.

실행:  HERMETIX_BROKER=next NEXT_CLIENT_ID=... python examples/larry.py
"""
from __future__ import annotations

import sys
from datetime import datetime, timedelta
from decimal import Decimal, ROUND_DOWN

sys.path.insert(0, __file__.rsplit("/", 2)[0])  # python/ 루트

from hermetix import Buy, CandleInterval, Sell, Signal, Strategy, StrategyContext, StrategySpec


class LarryStrategy(Strategy):

    def __init__(self, symbols: list[str] | None = None, lookback: int = 24,
                 multiplier: Decimal = Decimal("1.2"), expire_hours: int = 48,
                 budget_ratio: Decimal = Decimal("0.5"),
                 candle_interval: CandleInterval = CandleInterval.HOUR_1):
        self.symbols = symbols or ["AAPL"]
        self.lookback = lookback
        self.multiplier = multiplier
        self.expire_hours = expire_hours
        self.budget_ratio = budget_ratio
        self.spec = StrategySpec(
            name="larry", symbols=self.symbols,
            candle_interval=candle_interval, candle_limit=lookback + 2,
        )
        self._last_evaluated: dict[str, datetime] = {}
        self._entry_at: dict[str, datetime] = {}

    def decide(self, ctx: StrategyContext) -> list[Signal]:
        signals: list[Signal] = []
        for symbol in self.symbols:
            signals.extend(self._decide_symbol(symbol, ctx))
        return signals

    def _decide_symbol(self, symbol: str, ctx: StrategyContext) -> list[Signal]:
        if ctx.has_position(symbol):
            return self._decide_exit(symbol, ctx)
        self._entry_at.pop(symbol, None)
        if ctx.has_open_order(symbol):
            return []
        return self._decide_entry(symbol, ctx)

    def _decide_entry(self, symbol: str, ctx: StrategyContext) -> list[Signal]:
        candles = ctx.candles_of(symbol)
        if len(candles) < self.lookback + 2:
            return []
        completed = candles[:-1]           # 마지막 캔들은 진행 중
        target = completed[-1]
        if self._last_evaluated.get(symbol) == target.timestamp:
            return []                       # 같은 캔들로 중복 진입 방지
        self._last_evaluated[symbol] = target.timestamp

        history = completed[-(self.lookback + 1):-1]
        avg_body = sum(abs(c.open - c.close) for c in history) / len(history)
        if abs(target.open - target.close) < avg_body * self.multiplier:
            return []
        if target.open >= target.close:     # 롱 온리 - 음봉 돌파 스킵
            return []

        quote = ctx.quote(symbol)
        if quote is None:
            return []
        budget = ctx.buying_power * self.budget_ratio / len(self.symbols)
        quantity = (budget / quote.price).quantize(Decimal(1), rounding=ROUND_DOWN)
        if quantity < 1:
            return []

        self._entry_at[symbol] = ctx.now
        return [Buy(symbol=symbol, quantity=quantity, stop_loss_price=target.open)]

    def _decide_exit(self, symbol: str, ctx: StrategyContext) -> list[Signal]:
        opened_at = self._entry_at.setdefault(symbol, ctx.now)  # 재시작 시 만료 클록 재시작
        if ctx.now - opened_at < timedelta(hours=self.expire_hours):
            return []
        holding = ctx.holding(symbol)
        if holding is None:
            return []
        self._entry_at.pop(symbol, None)
        return [Sell(symbol=symbol, quantity=holding.quantity)]


if __name__ == "__main__":
    from _runner import run  # noqa: E402
    run(LarryStrategy())
